src/plot_api/plot_api.js

```diff
@@ -2508,6 +2527,360 @@ Plotly.relayout = function relayout(gd, astr, val) {
+        var hasAxisTransition = false;
+
+        if(layout) {
+            for(j = 0; j < basePlotModules.length; j++) {
+                if(basePlotModules[j].transitionAxes) {
+                    var newLayout = Lib.expandObjectPaths(layout);
+                    hasAxisTransition = basePlotModules[j].transitionAxes(gd, newLayout, transitionConfig, makeCallback) || hasAxisTransition;
+                }
+            }
+        }
+
+        // Here handle the exception that we refuse to animate scales and axes at the same
+        // time. In other words, if there's an axis transition, then set the data transition
+        // to instantaneous.
+        if(hasAxisTransition) {
+            traceTransitionConfig = Lib.extendFlat({}, transitionConfig);
+            traceTransitionConfig.duration = 0;
+        } else {
+            traceTransitionConfig = transitionConfig;
+        }
+
+        for(j = 0; j < basePlotModules.length; j++) {
+            // Note that we pass a callback to *create* the callback that must be invoked on completion.
+            // This is since not all traces know about transitions, so it greatly simplifies matters if
+            // the trace is responsible for creating a callback, if needed, and then executing it when
+            // the time is right.
+            basePlotModules[j].plot(gd, transitionedTraces, traceTransitionConfig, makeCallback);
+        }
+
+        // If nothing else creates a callback, then this will trigger the completion in the next tick:
+        setTimeout(makeCallback());
+
+        if(!hasAxisTransition && !hasTraceTransition) {
+            return false;
+        }
+
+    }
+
+    function completeTransition() {
+        flushCallbacks(gd._transitionData._interruptCallbacks);
+
+        return Promise.resolve().then(function() {
+            if(transitionConfig.redraw) {
+                return Plotly.redraw(gd);
+            }
+        }).then(function() {
+            gd.emit('plotly_endtransition', []);
+        });
+    }
+
+    function interruptPreviousTransitions() {
+        gd.emit('plotly_interrupttransition', []);
+
+        return executeCallbacks(gd._transitionData._interruptCallbacks);
+    }
+
+    for(i = 0; i < traceIndices.length; i++) {
+        traceIdx = traceIndices[i];
+        var contFull = gd._fullData[traceIdx];
+        var module = contFull._module;
+
+        if(!module) continue;
+
+        if(!module.animatable) {
+            var thisUpdate = {};
+
+            for(var ai in data[i]) {
+                thisUpdate[ai] = [data[i][ai]];
+            }
+
+            /*restyleList.push((function(md, data, traces) {
+                return function() {
+                    return Plotly.restyle(gd, data, traces);
+                };
+            }(module, thisUpdate, [traceIdx])));*/
+        }
+    }
+
+    var seq = [Plots.previousPromises, interruptPreviousTransitions, prepareTransitions, executeTransitions];
+    //seq = seq.concat(restyleList);
```

Comment:
is the restyle list made obsolete by the `Plotly.redraw` on the complete-transition callback?